File: Python/Easy/2273_Find_Resultant_Array_After_Removing_Anagrams/solution.py
class Solution:
    def removeAnagrams(self, words: list[str]) -> list[str]:
        output = [words[0]]

        for i in range(1, len(words)):
            if Solution.isAnagram(output[-1], words[i]):
                continue
            output.append(words[i])
        
        return output


    def isAnagram(s1, s2):
        if len(s1) != len(s2):
            return False
        
        s1Map = {}
        for w in s1:
            if w not in s1Map:
                s1Map[w] = 0
            s1Map[w] += 1
        
        for w in s2:
            if w not in s1Map:
                return False
            s1Map[w] -= 1
            if s1Map[w] == 0:
                del s1Map[w]

        return len(s1Map) == 0  

File: Golang/Easy/2273_Find_Resultant_Array_After_Removing_Anagrams/solution.go
package findresultantarrayafterremovinganagrams

func removeAnagrams(words []string) []string {
	output := []string{words[0]}

	for i := 1; i < len(words); i++ {
		if isAnagram(output[len(output)-1], words[i]) {
			continue
		}
		output = append(output, words[i])
	}
	return output
}

func isAnagram(s1 string, s2 string) bool {
	if len(s1) != len(s2) {
		return false
	}

	s1Map := make(map[rune]int)
	for _, w := range s1 {
		s1Map[w] += 1
	}

	for _, w := range s2 {
		if _, exist := s1Map[w]; !exist {
			return false
		}
		s1Map[w] -= 1
		if s1Map[w] == 0 {
			delete(s1Map, w)
		}
	}
	return len(s1Map) == 0
}
